executor/executor.go

```diff
@@ -122,22 +143,92 @@ func (e *Executor) Run(conflicts state.Keys, f func() error) {
 	e.tasks[id] = t
 
 	// Record dependencies
-	for k := range conflicts {
-		latest, ok := e.edges[k]
-		if ok {
-			lt := e.tasks[latest]
-			if !lt.executed {
-				if t.dependencies == nil {
-					t.dependencies = set.NewSet[int](defaultSetSize)
+	for k, v := range conflicts {
+		// Get last blocked transaction
+		parent, exists := e.edges[k]
+		if !exists {
+			concurrentReads := set.NewSet[int](defaultSetSize)
+			if v == state.Read {
+				// Record any reads on the same key to help
+				// with updating dependencies properly
+				concurrentReads.Add(id)
+			}
+			// Key doesn't exist, so we add it to our edge map
+			e.edges[k] = &KeyData{Parent: id, Permissions: v, ConcurrentReads: concurrentReads}
+			continue
+		}
+		pt := e.tasks[parent.Parent]
+		if !pt.executed {
+			if t.dependencies == nil {
+				t.dependencies = set.NewSet[int](defaultSetSize)
+			}
+			if pt.blocking == nil {
+				pt.blocking = set.NewSet[int](defaultSetSize)
+			}
+			// key has ONLY a Read permission
+			if v == state.Read {
+				parent.ConcurrentReads.Add(id)
```

Comment:
Could this lead to the same `id` being added to `pt.blocking` twice?

If `parent.Permissions == state.Write` and 1 tx comes in with `v==state.Read`, it will add here. If another tx comes in with `v==state.Write`, then it will iterate through the concurrent reads below and add the item to `pt.blocking`.

Reply:
reposting for visibility:

if i'm understanding this comment correctly, suppose we have the following 3 transactions:
T1 W
T2 R
T3 W

here, in T1 W, it's "parent" is T1. suppose it doesn't yet execute, for some reason

then in T2 R, it gets blocked by its "parent" which is T1, so we have the following:
T1.blocking = (T2)
T2.dep = (T1)
concurrentReads = (T2)

then when we run T3 W, we will go through concurrentReads and update the following:
T2.blocking = (T3)
T3.dep = (T2)

then we clear concurrentReads, and then finally, we update the "parent" to be T3

so in this walkthrough we don't add to pt.blocking twice. if a key has Allocate/Write, it will always be updated to itself